### PR TITLE
refactor: rm inventory emptyIds

### PR DESCRIPTION
**Pull request by noyyyy**
remove `inventoryEmptyIds`, no abi is changed. @ClaudeZsb merge it if approve

### CODE REVIEW

File: packages/contracts/src/library/Utils.sol [hunk 1]
```diff
@@ -46,13 +46,20 @@ library Utils {
...
       Player.updateInventory(_player, _index, uint64(0));
-      // add the index to empty Ids array
-      Player.pushInventoryEmptyIds(_player, uint8(_index));
     } else {
       revert("inv, out of index");
     }
   }
 
+  function getFirstInventoryEmptyIdx(address _player) internal view returns (uint256) {
+    for (uint256 i = 0; i < GameConfig.getInventorySlotNum(); i++) {
```
Comment (ClaudeZsb) [inline packages/contracts/src/library/Utils.sol:55]: I suggest that 
- Get out all values in inventory once. We now have 7 slots in inventory. This costs less gas than 2 times of gas for reading one slot. 
- Create a new local variable for inventory slot num. Set its value to length of inventory you got. Use it in the loop and save gas of reading one slot(`GameConfig.getInventorySlotNum()`) for each iteration.

Comment (ClaudeZsb) [reply packages/contracts/src/library/Utils.sol:55]: Why not keep on writing loop with `unchecked increment of i`. It's cheaper than `++i` which is cheaper than `i++`.

File: packages/contracts/src/systems/PlaceSystem.sol [hunk 2]
```diff
@@ -60,22 +60,16 @@ contract PlaceSystem is System {
    * @dev
    * @param index index of hero in Player.heroes
    */
-  function placeBackInventory(uint256 herosIndex) public onlyWhenGamePreparing returns (uint8 invIdx) {
+  function placeBackInventory(uint256 herosIndex) public onlyWhenGamePreparing returns (uint256 invIdx) {
```
Comment (ClaudeZsb) [inline packages/contracts/src/systems/PlaceSystem.sol:63]: What about adding a new input param `invIndex` which represents the index of slot in inventory where player wants to place the hero? I've seen it in a previous pr and think it's good.

Comment (noyyyy) [reply packages/contracts/src/systems/PlaceSystem.sol:63]: > What about adding a new input param `invIndex` which represents the index of slot in inventory where player wants to place the hero? I've seen it in a previous pr and think it's good.

There is a function `placeBackInventoryAndSwap` to do this, for backward compatibility. Will break this compatibility in another pr.

File: packages/contracts/src/library/Utils.sol (hunk 1 above)
Comment (noyyyy) [reply packages/contracts/src/library/Utils.sol:55]: u r right. will modify.

**Comment by noyyyy** [comment]
@ClaudeZsb please review again

File: packages/contracts/src/library/Utils.sol [hunk 3]
```diff
@@ -46,36 +46,42 @@ library Utils {
...
-      // add the index to empty Ids array
-      Player.pushInventoryEmptyIds(_player, uint8(_index));
     } else {
       revert("inv, out of index");
     }
   }
 
+  function getFirstInventoryEmptyIdx(address _player) internal view returns (uint256) {
+    uint64[] memory inv = Player.getInventory(_player);
+    uint256 length = GameConfig.getInventorySlotNum(0);
```
Comment (ClaudeZsb) [inline packages/contracts/src/library/Utils.sol:56]:
```
uint256 length = inv.length;
```

**Review by ClaudeZsb — CHANGES_REQUESTED** [review]
At `ShopSystem.sol:36`, getting an item from a dynamic array in mud store with an index larger than the array length will probably not produce a revert. And the value you got is not predictable.  So better remove a hero in altar in the same way as removing a hero in inventory.

File: packages/contracts/src/library/Utils.sol [hunk 4]
```diff
@@ -46,36 +46,42 @@ library Utils {
...
+    for (uint256 i = 0; i < length; i++) {
+      if (inv[i] == uint64(0)) {
+        return i;
+      }
+    }
+    revert("inventory full");
+  }
 
   function removeHeroByIndex(address _player, uint256 _index) internal returns (HeroData memory hero) {
```
Comment (ClaudeZsb) [inline packages/contracts/src/library/Utils.sol:65]: Those two funcs `removeHeroByIndex` and `deleteHeroByIndex` do the same thing but in different ways. @aLIEzsss4 
Do we need to maintain the order of hero in `Player.heroes` array?  If the answer is yes, then we keep `removeHeroByIndex` and remove the other one. If not, do the reverse.

File: packages/contracts/src/library/Utils.sol (hunk 3 above)
Comment (noyyyy) [reply packages/contracts/src/library/Utils.sol]: > ```
> uint256 length = inv.length;
> ```

Fixed

**Comment by noyyyy** [comment]
> At `ShopSystem.sol:36`, getting an item from a dynamic array in mud store with an index larger than the array length will probably not produce a revert. And the value you got is not predictable. So better remove a hero in altar in the same way as removing a hero in inventory.

If index is larger than length -1, the value will be zero, and we have zero check to make this revert.
`require(creatureId != 0, "empty hero altar slot");`

File: packages/contracts/src/library/Utils.sol (hunk 4 above)
Comment (noyyyy) [reply packages/contracts/src/library/Utils.sol:65]: > `removeHeroByIndex`

No. They are not the same. `removeHeroByIndex` delete hero in heros array, `deleteHeroByIndex` delete hero in inventory array.

Comment (noyyyy) [reply packages/contracts/src/library/Utils.sol:65]: > Those two funcs `removeHeroByIndex` and `deleteHeroByIndex` do the same thing but in different ways. @aLIEzsss4 Do we need to maintain the order of hero in `Player.heroes` array? If the answer is yes, then we keep `removeHeroByIndex` and remove the other one. If not, do the reverse.

`removeHeroByIndex` is not used. So I think @aLIEzsss4 has make his decision.

**Comment by ClaudeZsb** [comment]
> > At `ShopSystem.sol:36`, getting an item from a dynamic array in mud store with an index larger than the array length will probably not produce a revert. And the value you got is not predictable. So better remove a hero in altar in the same way as removing a hero in inventory.
> 
> If index is larger than length -1, the value will be zero, and we have zero check to make this revert. `require(creatureId != 0, "empty hero altar slot");`

The reason why I said ***The value you got is not predictable*** is that the index(`uint256`) could be extremely large. And `Mud` do not check the index for us. They just calculate and read from the start to the end slot number. In the worst case, a malicious player could read at most next `1/4 * type(uint256).max` slots and we have no idea now what will be stored there. So I recommend that we do something to avoid any reading overflowed into other table or dynamic field.

**Comment by noyyyy** [comment]
> 

I see. But this will not happen since we set inventory to a fixed length array when refresh altar.

https://github.com/HelheimLabs/autochessia/blob/f5fe8a8f2d3f6c93c5b5e3abe96ef1f0a64be936/packages/contracts/src/systems/RefreshHeroesSystem.sol#L14-L56


If the index is larger than array length. contract will revert. We can add a test to make sure of it.

**Comment by noyyyy** [comment]
> > > At `ShopSystem.sol:36`, getting an item from a dynamic array in mud store with an index larger than the array length will probably not produce a revert. And the value you got is not predictable. So better remove a hero in altar in the same way as removing a hero in inventory.
> > 
> > 
> > If index is larger than length -1, the value will be zero, and we have zero check to make this revert. `require(creatureId != 0, "empty hero altar slot");`
> 
> The reason why I said _**The value you got is not predictable**_ is that the index(`uint256`) could be extremely large. And `Mud` do not check the index for us. They just calculate and read from the start to the end slot number. In the worst case, a malicious player could read at most next `1/4 * type(uint256).max` slots and we have no idea now what will be stored there. So I recommend that we do something to avoid any reading overflowed into other table or dynamic field.


I see. But this will not happen since we set inventory to a fixed length array when refresh altar.

https://github.com/HelheimLabs/autochessia/blob/f5fe8a8f2d3f6c93c5b5e3abe96ef1f0a64be936/packages/contracts/src/systems/RefreshHeroesSystem.sol#L14-L56


If the index is larger than array length. contract will revert. We can add a test to make sure of it.

**Comment by noyyyy** [comment]
> The reason why I said _**The value you got is not predictable**_ is that the index(`uint256`) could be extremely large. And `Mud` do not check the index for us. They just calculate and read from the start to the end slot number. In the worst case, a malicious player could read at most next `1/4 * type(uint256).max` slots and we have no idea now what will be stored there. So I recommend that we do something to avoid any reading overflowed into other table or dynamic field.

Got it. u r right.

**Review by ClaudeZsb — APPROVED** [review]

**Pull request merged**